model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -266,6 +266,40 @@ message Metrics {
       map<string, int64> output_elements_remaining = 3;
     }
 
+    // User defined metrics
+    message User {
+
+      // A key for identifying a metric at the most granular level.
+      message MetricName {
+        // (Required): The namespace of this metric.
+        string namespace = 2;
+
+        // (Required): The name of this metric.
+        string name = 3;
+      }
+
+      // Data associated with a counter metric.
+      message CounterData {
```

Comment:
Incidentally, these make sense at the top level and I have a slight preference for keeping as many things toplevel as possible unless these is a compelling reason to do otherwise. There is usually no downside.

Here it is a particular readability problem:

 - `User` is very misleading if not fully qualified, versus e.g. `MetricData`
 -  great distance between the start of the message and the fields

Reply:
(and is `MetricData` the same as the existing `MetricUpdate`? Luke and I chatted and I came away unsure)

Reply:
Moving stuff to not be nested is an independent change to what was asked. klk@, if you would like to take on that change, feel free to do so.

Reply:
I might, depending on how much this proto enters my life. But we should fix the misleading name either way.

Reply:
BTW I don't know of a public discussion where I could gain context on this change. Can you link to that and/or public design doc on user-defined metrics in portability?

Anyhow I'm just providing drive-by feedback on the protos since I tried to use them and it isn't really clear why they are the way they are. I'm sure I can power through and make it work, but it isn't great for people not in our office.